src/fu-engine.c

```diff
@@ -2223,19 +2227,52 @@ fu_engine_install_releases(FuEngine *self,
 		if (!fu_engine_install_release(self,
 					       release,
 					       stream,
-					       fu_progress_get_child(progress),
+					       child,
 					       flags,
-					       error)) {
+					       &error_local_install)) {
+			FuDevice *device = fu_release_get_device(release);
 			g_autoptr(GError) error_local = NULL;
-			if (!fu_engine_composite_cleanup(self, devices, &error_local)) {
-				g_warning("failed to cleanup failed composite action: %s",
-					  error_local->message);
+
+			/* resume composite updates on device error */
+			if (fu_device_has_private_flag(
+				device,
+				FU_DEVICE_PRIVATE_FLAG_COMPOSITE_ERROR_CONTINUE)) {
+				/* terminate the child progress */
+				fu_progress_finished(child);
+
+				/* store the update error message */
+				g_string_append_printf(composite_error_message,
+						       "failed composite update on %s: %s. ",
+						       fu_device_get_name(device),
+						       error_local_install->message);
+			} else {
+				if (!fu_engine_composite_cleanup(self, devices, &error_local)) {
+					g_warning("failed to cleanup failed composite action: %s",
+						  error_local->message);
+				}
+				return FALSE;
 			}
-			return FALSE;
 		}
 		fu_progress_step_done(progress);
 	}
 
+	/* abort because an error already occurred */
+	if (composite_error_message->len > 0) {
+		g_autoptr(GError) error_local = NULL;
+
+		/* best effort to cleanup the devices */
+		if (!fu_engine_composite_cleanup(self, devices, &error_local))
+			g_warning("failed to cleanup failed composite action: %s",
+				  error_local->message);
+
+		/* propagate the error message */
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    composite_error_message->str);
```

Comment:
The error could be generated from one or two (or more) out of the composite devices, and it could be any dynamic type of error that can't be predicted.

The error should be managed to a specific error type, so we the process failed in this way.